pkg/io/video/copy.go

```diff
@@ -0,0 +1,196 @@
+package video
+
+import (
+	"image"
+)
+
+// Copier is an optimized image copier that tries to reduce the number of memory allocations
+// by reusing the memory area that holds pixels.
+type Copier struct {
+	buffer []uint8
+}
+
+// NewCopier creates a new copier instance and initialize internal buffer with initialSize.
+func NewCopier(initialSize int) *Copier {
+	return &Copier{
+		buffer: make([]uint8, initialSize),
+	}
+}
+
+func (copier *Copier) storeInOrder(srcs ...[]uint8) {
+	var neededSize int
+
+	for _, src := range srcs {
+		neededSize += len(src)
+	}
+
+	if len(copier.buffer) < neededSize {
+		if cap(copier.buffer) >= neededSize {
+			copier.buffer = copier.buffer[:neededSize]
+		} else {
+			copier.buffer = make([]uint8, neededSize)
+		}
+	}
+
+	var currentLen int
+	for _, src := range srcs {
+		copy(copier.buffer[currentLen:], src)
+		currentLen += len(src)
+	}
+}
+
+// Copy copies src to dst. Since dst can be nil, the caller is responsible to assign the
+// return value from Copy to dst. When dst is not nil, Copy will reuse its memory and only
+// copy values from src to dst.
```

Comment:
I think it's better to split shallow copy func and deep copy func with explicit names.

Reply:
Not sure what you mean, but this function only does deep copy.

Reply:
> When dst is not nil, Copy will reuse its memory and only copy values from src to dst.

This function looks doing deep copy only if dst is not nil.

Reply:
Oh I see what you mean. So, I guess you're thinking about users misusing this by calling copy from the same Copier instance in different goroutines. I can definitely see that to be problematic.

Reply:
So, I think we'll create `ShallowCopy` and `DeepCopy`. But, should we allow `ShallowCopy` to have `nil` as its dst? Or should we just enforce users to not have `nil` at all as its dst?